application_logs: add bootstrap option to write logs in JSON format

ohadvano opened the pull request:
Commit Message: application_logs: add bootstrap option to write logs in JSON format
Additional Description: Adds an option in bootstrap config to write application logs in JSON format, while supporting all the log-format flags as defined in the CLI ``--log-format`` option. Related to #25959 - this is the first step in the implementation for supporting custom JSON properties, while printing the application logs output in JSON format.
Risk Level: Low (all new code paths are only enabled by config option)
Testing: Unit tests
Docs Changes: API, Application logs docs
Release Notes: None
Platform Specific Features: None

ohadvano commented:
cc @mattklein123 as you engaged in the issue thread, will appreciate your thoughts on this, thanks.

Example usage:

```
application_log_format:
  json_format:
    ProcessId: "%P"
    Message: "%v"
    Level: "%l"
    Time: "%Y-%m-%d %T.%e"
```

Expected output:
`{"Time":"2023-05-09T14:03:50.753171170","ProcessId":"1010","Level":"info","Message":"my_message"}`

ohadvano commented:
I added implementation based on the first approach (see previous comment). Will appreciate initial review before proceeding to unit tests, to make sure this is on track.

/assign-from @envoyproxy/senior-maintainers

adisuissa commented on source/server/server.cc:
```diff
@@ -425,6 +425,11 @@ void InstanceImpl::initialize(Network::Address::InstanceConstSharedPtr local_add
 
+  if (bootstrap_.has_application_log_format() &&
+      bootstrap_.application_log_format().has_json_format()) {
+    Logger::Registry::setJsonLogFormat(bootstrap_.application_log_format().json_format());
```
should probably be added to the config_validation as well.

adisuissa replied:
There's a separate server just for config-validation:
https://github.com/envoyproxy/envoy/blob/b978af64641aa46d3b7012ec1d9f3d7d9fd3b307/source/server/config_validation/server.cc#L72

ohadvano replied:
gotcha, added. Thanks.

adisuissa commented on source/common/common/logger.cc:
```diff
@@ -254,6 +256,22 @@ void Registry::setLogFormat(const std::string& log_format) {
+  if (auto status =
+          Protobuf::util::MessageToJsonString(log_format_struct, &format_as_json, json_options);
+      !status.ok()) {
+    throw EnvoyException(
```
I'm pretty certain this won't be used in the data-plane code, but just to be safe here, can this function return absl::Status, and the exception will be thrown in the server's code?

ohadvano replied:
done

adisuissa reviewed (commented):
Thanks!
At a high level API looks good other than a few nits.
Left a few minor code comments.

ohadvano commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -101,6 +101,17 @@ message Bootstrap {
+
+      // Flush application logs in JSON format. The configured JSON struct can
+      // support all the format tags specified in the --log-format section,
+      // in :ref:`operations CLI <operations_cli>`.
```
done

ohadvano commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -360,6 +371,10 @@ message Bootstrap {
   core.v3.TypedExtensionConfig listener_manager = 37;
+
+  // Optional field to set the application logs format. If this field is set, it will override
+  // the default log format, or the format that is specified by the CLI option ``--log-foramt``.
```
done

ohadvano replied:
Yeah I understand your point. I'm not familiar with other CLI options that also have a similar configurable feature, do you?
I think there's a good point for this override, specifically because if you want to support full JSON experience then you would need to use both ``--log-format`` as well as the ``application_log_format`` - I elaborated the full details in the ``application_logs`` docs. Wdyt?

ohadvano commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -101,6 +101,15 @@ message Bootstrap {
+    oneof log_format {
+      // Flush application logs in JSON format. If set, it will override Envoy's --log-format
+      // CLI option. The configured JSON struct can support all the format tags specified in
+      // the --log-format section, in :ref:`operations CLI <operations_cli>`.
```
done

ohadvano commented:
/coverage

adisuissa commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -360,6 +371,10 @@ message Bootstrap {
   core.v3.TypedExtensionConfig listener_manager = 37;
+
+  // Optional field to set the application logs format. If this field is set, it will override
+  // the default log format, or the format that is specified by the :option:`--log-format` command line option.
```
Looking at the `--service-node` option and it relation to [Node.id](https://github.com/envoyproxy/envoy/blob/5e7a4b5f44f398a42897f743a8665ce92fc563a3/api/envoy/config/core/v3/base.proto#L156), it seems that the CLI argument overrides the bootstrap config ([relevant test](https://github.com/envoyproxy/envoy/blob/5e7a4b5f44f398a42897f743a8665ce92fc563a3/test/server/server_test.cc#L1026)). 
I suggest to keep it consistent, i.e., CLI overrides Bootstrap overrides default value.

Please also add a test that validates that the correct precedence is kept.

ohadvano replied:
Thanks for finding another exmple. However, did you look at the use case, I worte in the application logs doc? If the config doesn't override the CLI option, this won't be available to achieve. I think that it's good enough reason to keep this order of precendance, even at the cost of inconsistantcy with other options precedance..

snowp replied:
Not sure I follow - why can't we have the bootstrap config be used when `--log-format` is not specified by the command line?

ohadvano replied:
I'm not saying that it technically can't be, but the problem is this: since the bootstrap config option is loaded only after some logs are written (the initialization logs) it causes inconsistency, meaning that initialization logs will be printed in default format, and the rest of the logs will be written in the bootstrap configured format.

However, this can be mitigated, if the user configures both ``--log-format`` in manual JSON format, and the bootstrap config. This way, initialization logs will be written by the ``--log-format`` JSON format config, and the rest will be by the bootstrap JSON config.

If we let the CLI option override, then this mitigation is not possible. If you both think that this mitigation isn't important then we can give this up and let the CLI config a precedence. (For me, the initialization logs don't help too much, so I wob't insist on this..). 

Please lmk what you think @adisuissa, @snowp

ohadvano replied:
BTW, implementation-wise, it doesn't seem straightforward, because ``options.logFormat()`` gets a default value, so there needs to be a way at the server context to know if this was set in the first place, and this would require some refactoring.

adisuissa replied:
I think there will always be cases of "logging" before the log-format (even the one provided using the command line) is set. For example, IIUC [this](https://github.com/envoyproxy/envoy/blob/dc860e850c577484d31b7b7e3697d6de95cc6d01/source/exe/stripped_main_base.cc#L60) can be printed before the log-format is being used.

From a user's experience point-of-view, I still think it makes sense that the command line (if given) overrides whatever is set in the bootstrap (see #2269). The way I would expect things this to work in practice is that there's some bootstrap config that is typically used, and if the operator wants to override it, they add a CLI argument (similar to [overriding the node ID](https://www.envoyproxy.io/docs/envoy/latest/operations/cli.html#cmdoption-config-yaml)). If possible, it may be beneficial to keep this behavior.

That said, I don't feel strongly about this. cc @envoyproxy/envoy-maintainers for other ideas.

ohadvano replied:
Got it, I think we can stick to this behavior and let it override

alyssawilk reviewed (changes requested):
please make sure you add tests covering all new code, rather than lowering coverage limits.

ohadvano commented:
@alyssawilk, the uncovered lines are the ``throw`` statements in the server.cc code, however I'm not sure if they are reachable, because the proto serializer will enforce the construction of ``google.protobuf.Struct`` object. @adisuissa to make this testable I'll have to revert your suggestion so the ``throw`` statement will go back to the ``logger`` code.

ohadvano commented:
I did add tests for ``ValidationInstance::Initialize``/``Instance::Initialize (server)``, however I suspect that some of the lines added to ``config_validation.cc``/``server.cc`` are not reachable: 

https://github.com/envoyproxy/envoy/blob/4a2495ad3554e3cab9110bc835e7bd04788f94b9/source/server/config_validation/server.cc#L92

This is because they are tested with actual yaml config, and I'm not sure if it's possible to compose such yaml that will cause the ``status`` returned to be ``ok() == false``, because if the yaml was de-serialized, we get ``google.protobuf.Struct`` which will in any way return ``ok() == true`` in the ``setJsonLogFormat`` implementation. Otherwise, can you please suggest a yaml config that would cause the returned status to

alyssawilk commented:
which test attempts to test with invalid data?
alternately if you're adding code which 100% can not be tested you're welcome to find a different part of the file to add unit tests for.
/wait

ohadvano commented:
I'll resolve this in the next iteration. Awaiting @adisuissa and @snowp for response about the precedence topic (friendly ping)

adisuissa commented on docs/root/configuration/observability/application_logging.rst:
```diff
@@ -23,3 +23,27 @@ with the following :ref:`command line options <operations_cli>`:
+  Therefore, some of the logs will be printed in the default log format, or the one that is specified by the CLI option ``--log-format``.
+  For full JSON application logs experience, it is possible to set ``--log-format`` with JSON format,
+  so both logs before bootstrap initialization and after, are written in JSON format.
+  For example: ``--log-format '{"Timestamp":"%Y-%m-%dT%T.%F","ThreadId":"%t","SourceLine":"%s:%#","Level":"%l","Message":"%_"}'``.
```
Should the `%_` be replaced with `%j`?

adisuissa commented on source/common/common/logger.cc:
```diff
@@ -254,6 +254,21 @@ void Registry::setLogFormat(const std::string& log_format) {
+  auto status =
+      Protobuf::util::MessageToJsonString(log_format_struct, &format_as_json, json_options);
+  if (status.ok()) {
+    setLogFormat(format_as_json);
```
I think the logged message may not be a valid JSON escaped string. That is, the pattern will be, but the logged message may not.
Will it make sense to replace/reject a config with `%_`?

ohadvano replied:
Sorry but I'm not sure if I understand what you mean. "%_" stands for "The actual message to log, but with escaped newlines". If the schema log schema is:

```
application_log_format:
  json_format:
    Message: "%_"
```

Wouldn't the outcome be `{"Message":"message that was escaped\nhere"}`?

Did you (maybe) mean that we should enforce that "%v" is not used? Because %v stands for "The actual message to log", so for valid JSON format the message couldn't have newlines in it?

ohadvano commented:
Thanks @adisuissa for the allowed flags insight, it was correct - ``%v`` can't be used, because if there's some log, for example ``ENVOY_LOG(info, "hel\nlo")``, it would break the JSON output. Therefore I disallowed it and added tests to cover this case. ``%_`` and ``%j`` however, are OK, because they are escaping such characters and therefore the output would be a one-liner message, added tests to verify these as well and updated docs to note about it. Will appreciate review for latest iteration.

adisuissa commented on source/server/config_validation/server.cc:
```diff
@@ -85,6 +85,16 @@ void ValidationInstance::initialize(const Options& options,
 
+  if (!options.logFormatSet() && bootstrap_.has_application_log_format() &&
+      bootstrap_.application_log_format().has_json_format()) {
+    auto status =
```
nit: const

ohadvano replied:
done

adisuissa commented on source/common/common/logger.cc:
```diff
@@ -254,6 +254,27 @@ void Registry::setLogFormat(const std::string& log_format) {
+  json_options.always_print_primitive_fields = true;
+
+  std::string format_as_json;
+  auto status =
```
nit: const

ohadvano replied:
done

adisuissa commented on test/server/config_validation/server_test.cc:
```diff
@@ -206,6 +237,70 @@ INSTANTIATE_TEST_SUITE_P(
+  Envoy::Logger::Registry::setLogLevel(spdlog::level::info);
+  MockLogSink sink(Envoy::Logger::Registry::getSink());
+  EXPECT_CALL(sink, log(_, _)).WillOnce(Invoke([](auto msg, auto& log) {
+    EXPECT_THAT(msg, HasSubstr("{\"MessageFromProto\":\"hello\"}"));
```
Consider parsing the output as a JSON object (to prevent future changes that may add something bad that breaks the JSON log).

ohadvano replied:
done

adisuissa commented on test/server/config_validation/server_test.cc:
```diff
@@ -73,14 +77,41 @@ class RuntimeFeatureValidationServerTest : public ValidationServerTest {
-    }
-    return files;
+  static void setupTestDirectory() {
+    directory_ =
```
I may be missing something here, but why is the `directory_` needed?

ohadvano replied:
It is used by the MockOptions object maintained in the ``ValidationServerTest`` above - that's the source directory that the yaml configs are loaded for the tests. It was already embedded in the current test pattern (before I added this change), so I followed the same pattern.

adisuissa reviewed (commented):
Thanks!
Overall makes sense to me. Left a few minor comments.
/lgtm api

/assign-from @envoyproxy/senior-maintainers

ohadvano commented on test/common/common/logger_test.cc:
```diff
@@ -270,6 +260,109 @@ TEST(LoggerTest, LogWithLogDetails) {
+  EXPECT_CALL(sink, log(_, _))
+      .WillOnce(Invoke([](auto msg, auto& log) {
+        EXPECT_NO_THROW(Json::Factory::loadFromString(std::string(msg)));
+        EXPECT_THAT(msg, HasSubstr("\"Level\":\"info\""));
+        EXPECT_THAT(msg, HasSubstr("\"Message\":\"hello\""));
+        EXPECT_THAT(msg, HasSubstr("\"FixedValue\":\"Fixed\""));
```
I did try this, but something is off with asserting the JSON itself and I couldn't figure out what, so I preferred to stay with string checks for the integrity of the tests.

For example, if I take those 3 lines and change them to the following, then the test will pass, even though there are obviously no 1000 objects in the JSON object and there's no such value:

```
auto json = Json::Factory::loadFromString(std::string(msg));
EXPECT_EQ(1000, json->asObjectArray().size());
EXPECT_EQ("no-such-value", json->getString("Message"));
```

I tried throwing some random exception in the test to see if something is written in the test output, and there's this line:
`[*** LOG ERROR #0001 ***] [2023-05-22 16:32:44] [misc] {JSON field from line 1 accessed with type 'Array' does not match actual type 'Object'. [test/common/common/logger_test.cc(288)]}`

ohadvano commented:
/retest

ohadvano commented:
Ping @snowp

ohadvano commented:
Drafting again because I figured out that the quote mark in a message will also mess the JSON format. Will open back after fixing this.

ohadvano commented:
@adisuissa I edited the API comment so the api tag got back, will appreciate if you can review again.

ohadvano commented:
/retest

phlax commented:
/wait-any for further review

cc @adisuissa @alyssawilk

alyssawilk reviewed (commented):
some high level comments before I look at code
/wait

ohadvano commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -360,6 +371,11 @@ message Bootstrap {
+  // Optional field to set the application logs format. If this field is set, it will override
+  // the default log format. In case the :option:`--log-format` command line option is used,
+  // it will override ``application_log_format`` configurations.
+  ApplicationLogFormat application_log_format = 38;
```
The ``application_log_format`` would now support json format, but it can be extended to more types of formats by adding more options to the ``ApplicationLogFormat`` message - this is why I used ``oneof`` to allow extensibility of more types of formats

ohadvano commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -360,6 +371,11 @@ message Bootstrap {
+
+  // Optional field to set the application logs format. If this field is set, it will override
+  // the default log format. In case the :option:`--log-format` command line option is used,
+  // it will override ``application_log_format`` configurations.
```
Summarizing previous discussion on this PR - @adisuissa suggested to keep this behavior consistent with how ``--service-node`` behaves (initially I let precedence for bootstrap config, but changed it after Adi's insights) where if there are both ``--service-node`` and bootstrap ``node`` set, then CLI option has precedence, but do not throw an error.

alyssawilk replied:
@adisuissa is out.  tagging @mattklein123 for second opinion.  I prefer explicit failures to silently ignoring config.

Also above, I don't love having a new bootstrap option that supercedes but doesn't support the existing log modes.  Would like a second opinion on that.

ohadvano replied:
So the gap here is supporting a freestyle ``text_format`` option in addition to the ``json_format`` in ``ApplicationLogFormat``?
I don't object to adding that, and I think it makes sense to have it - I can take this in another PR.

mattklein123 replied:
The node ID stuff is the way it is due to (very, very) long legacy. I would't do that again. I agree with @alyssawilk to make this a failure and not silently override.

Also agree that we should support what the CLI supports as well. Can we open a tracking issue on that and if you are willing to take it that would be great.

ohadvano replied:
Opened an issue for tracking: https://github.com/envoyproxy/envoy/issues/27761

alyssawilk commented:
@mattklein123 ping ?

mattklein123 commented on docs/root/configuration/observability/application_logging.rst:
```diff
@@ -23,3 +23,24 @@ with the following :ref:`command line options <operations_cli>`:
+      FixedValue: "SomeFixedValue"
+
+.. note::
+  In case the CLI option ``--log-format`` is used, its value will override ``application_log_format`` format.
```
Per above let's change this.

ohadvano replied:
done

mattklein123 commented on docs/root/configuration/observability/application_logging.rst:
```diff
@@ -23,3 +23,24 @@ with the following :ref:`command line options <operations_cli>`:
 
 `Reference documentation <https://cloud.google.com/run/docs/logging#container-logs>`_ for Stackdriver on GKE.
+
+Printing logs in JSON format
```
This feature needs a release note.

mattklein123 reviewed (changes requested):
Cool feature. Some high level comments, thank you!

/wait

ohadvano commented:
I disallowed both configs to be set, and opened an issue to track bootstrap textual format feature: https://github.com/envoyproxy/envoy/issues/27761
I'll get to it right after this one is merged

mattklein123 commented on source/server/config_validation/server.cc:
```diff
@@ -85,7 +85,12 @@ void ValidationInstance::initialize(const Options& options,
                                     messageValidationContext().staticValidationVisitor(), *api_);
 
-  if (!options.logFormatSet() && bootstrap_.has_application_log_format() &&
+  if (options_.logFormatSet() && bootstrap_.has_application_log_format()) {
+    throw EnvoyException(
+        "Only one of application_log_format or CLI option --log-format can be specified.");
+  }
```
nit: I would factor here and below into some tiny shared helper just so we share the same logic and error message in case it changes in the future.

ohadvano replied:
done

mattklein123 reviewed (changes requested):
LGTM with small comment, thanks.

/wait

tsaarni commented on api/envoy/config/bootstrap/v3/bootstrap.proto:
```diff
@@ -101,6 +101,17 @@ message Bootstrap {
     core.v3.ApiConfigSource ads_config = 3;
   }
 
+  message ApplicationLogFormat {
```
Would it make sense to rename the message to something more generic, like `AppicationLog`? 

There might be need for other application log related configuration in the (near) future. I've been working on https://github.com/envoyproxy/envoy/pull/27579 (design doc to-be-written) and I likely need something in bootstrap to configure application log masking rules. So maybe it would be nice to have common message to avoid adding many top-level application log related fields?

Cc @yanavlasov

ohadvano replied:
Ack - I nested the log format config in ApplicationLogConfig message. Thanks for that masking work - seems important.

mattklein123 reviewed (dismissed):
Awesome, thanks!

mattklein123 commented:
LGTM. Not sure if @alyssawilk has any other comments so I will let her merge.

alyssawilk reviewed (commented):
thanks for swapping to disallow both configs

I'm mildly concerned this PR is going to silently break the mobile build due to CI being set up correctly - there's a new build option on the E-M side I think this PR will tickle, and I'm not seeing the tests run.

I'm trying to fix upstream CI to run correctly  over in https://github.com/envoyproxy/envoy/pull/27707 but as an interim workaround if you don't want to wait, you could do the horrible hacky workaround of editing a file (reword a comment in mobile/test/common/integration/) and see if this breaks the mobile compile time options build.  sorry for the hassle but it beats a revert!

/wait

ohadvano commented:
@alyssawilk I'm completely clueless about mobile build/tests - there's an iOS integration test failing. Is it related to the PR changes or may it be transient?

ohadvano commented:
/retest

alyssawilk commented:
very likely a flake - retest will verify.  a few of the mobile CI builds especially the ones that pull external dependencies have network errors, which is one of the reasons we don't run them on all CI :-)
the only one I was concerned about was cc_tests in compile time options so the PR is in the submit queue and retest should land it CI willing :-)

ohadvano commented:
/retest

phlax commented:
@alyssawilk @ohadvano this failed clang-tidy and was merged anyway afaict https://dev.azure.com/cncf/envoy/_build/results?buildId=139360&view=logs&j=b7634614-24f3-5416-e791-4f3affaaed6c&t=d565ee52-b888-5d80-7fea-dfaed104c390&l=167

ohadvano commented:
@phlax nope, all CI steps passed: https://dev.azure.com/cncf/envoy/_build/results?buildId=139347&view=results
Any chance the clang-tidy checks were changed recently and wasn't rebased into the PR branch?

phlax commented:
>  Any chance the clang-tidy checks were changed recently and wasn't rebased into the PR branch?

the clang tidy test is not great - it uses a diff, takes ages, and seems to miss stuff so its quite possible that for some reason it didnt detect the change correctly and report the issue

either way it failed on postsubmit

it also triggered an asan issue which i was just trying to track and cant see any previous occurrence https://dev.azure.com/cncf/envoy/_build/results?buildId=139360&view=logs&j=1439b9f7-a348-5b50-b5fe-ea612ea91241&t=20bd2bcd-3fb5-5bf3-9cdb-ca6f94dd3096&l=2209 - altho this test flakes pretty regularly so not sure at all about that

in the case of the asan test im guessing its just triggered something #27814
